client/rpc.go

```diff
@@ -4,70 +4,260 @@ package client
+	// Setup async response cb.
+	r.registerCallback(call.Call.ID, func(result wire.Result, msg []byte) {
+		if result.Error != "" {
+			errChan <- fmt.Errorf("SendTx RPC result: %s", result.Error)
+		} else {
+			errChan <- nil
+		}
+	})
+	// Make the call.
+	if err := r.sendJSON(call); err != nil {
+		return fmt.Errorf("sending json object: %w", err)
+	}
+	// Return error from async response cb.
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
 }
 
-func (r *RPC) GetDepositProof(ctx context.Context, epoch uint64, user common.Address) (reply tee.DepositProof, err error) {
-	for i := 0; ; i++ {
-		if e := r.call(ctx, "RemoteEnclave.GetDepositProof", user, &reply); e != nil {
-			return tee.DepositProof{}, e
+// Subscribe subscribes to the Balance and Deposit proof topic.
+// The user must always read the proofs from via `DepositProof` and
+// `BalanceProof`.
+func (r *RPC) Subscribe(ctx context.Context, user common.Address) (*Subscription, error) {
+	r.subMtx.Lock()
+	r.subscription = &Subscription{
+		Embedding: perunlog.MakeEmbedding(perunlog.WithField("role", "proofSub")),
+		// Buffer the proofs here, otherwise the client has to read them
+		// immediately to prevent that they get reordered by a race condition
+		// from the go-routines writing them to the channel since a mutex is
+		// not FIFO.
+		balProofs: make(chan tee.BalanceProof, 10),
+		depProofs: make(chan tee.DepositProof, 10),
+	}
+	r.subMtx.Unlock()
+
+	call := wire.NewSubscribe(r.nextID(), user)
+	errChan := make(chan error)
+	// Setup async response cb.
+	r.registerCallback(call.Call.ID, func(result wire.Result, msg []byte) {
+		if result.Error != "" {
+			errChan <- fmt.Errorf("Subscribe RPC result: %s", result.Error)
+		} else {
+			errChan <- nil
+		}
+	})
+	// Make the call.
+	if err := r.sendJSON(call); err != nil {
+		return nil, fmt.Errorf("sending json object: %w", err)
+	}
+	// Return error from async response cb.
+	select {
+	case err := <-errChan:
+		return r.subscription, err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (r *RPC) handleConnections() {
+	for !r.IsClosed() {
+		// gorilla has no async read method?!
+		_, data, err := r.conn.ReadMessage()
+		if err != nil {
+			r.Log().Error("reading ws message: ", err)
+			time.Sleep(1 * time.Second)
+			continue
 		}
-		if reply.Balance.Epoch < epoch {
-			time.Sleep(time.Second * 1)
+		r.Log().Trace("client received: ", string(data))
+		var msg wire.Result
+		if err := json.Unmarshal(data, &msg); err != nil {
+			r.Log().Error("decoding message: ", err)
 			continue
-		} else if reply.Balance.Epoch > epoch {
-			return tee.DepositProof{}, errors.New("Newer epoch")
 		}
+
+		switch {
+		case msg.ID != "":
+			r.callCallback(msg, data)
+		case msg.Topic != "":
+			r.subMtx.RLock()
+			// Don't unlock here, since otherwise a `Subscribe` call could
+			// override the subscription while it receives a proof.
+			defer r.subMtx.RUnlock()
```

Comment:
Never call `defer`s inside loops. What you're doing here is acquiring indefinitely many read locks and stack indefinitely many unlocks, all executed at once when the `handleConnections` go routine returns.

Also I'm not sure if we even need the `subMtx`. I'll leave a comment about this somewhere else.

Reply:
🤦